anonymizer/OpenDiffix.Service.Tests/CSVFormatter.Tests.fs

```diff
@@ -0,0 +1,82 @@
+module OpenDiffix.Service.CSVFormatterTests
+
+open Xunit
+open FsUnit.Xunit
+
+open OpenDiffix.Core
+open OpenDiffix.Core.QueryEngine
+
+[<Fact>]
+let ``Formats result with star bucket`` () =
+  let result =
+    {
+      Columns =
+        [
+          { Name = "age"; Type = IntegerType }
+          { Name = "city"; Type = StringType }
+          { Name = "discount"; Type = RealType }
+          { Name = "active"; Type = BooleanType }
+          { Name = "count"; Type = IntegerType }
+        ]
+      Rows = [ [| Integer 18L; String "Berlin"; Real 0.5; Boolean true; Integer 1L |] ]
+    }
+
+  let suppressedAnonCount = Integer 2L
+  let CSVResult = (CSVFormatter.format suppressedAnonCount result).Split('\n')
+
+  CSVResult.[0]
+  |> should equal "\"age\",\"city\",\"discount\",\"active\",\"count\""
+
+  CSVResult.[1] |> should equal "\"*\",\"*\",\"*\",\"*\",2"
+  CSVResult.[2] |> should equal "18,\"Berlin\",0.5,True,1"
+  CSVResult.Length |> should equal 3
+
+[<Fact>]
+let ``Formats result without the star bucket`` () =
+  let result =
+    {
+      Columns = [ { Name = "count"; Type = IntegerType } ]
+      Rows = [ [| Integer 1L |] ]
+    }
+
+  let suppressedAnonCount = Null
+  let CSVResult = (CSVFormatter.format suppressedAnonCount result).Split('\n')
+
+  CSVResult.[0] |> should equal "\"count\""
+  CSVResult.[1] |> should equal "1"
+  CSVResult.Length |> should equal 2
+
+[<Fact>]
+let ``Handles quotes`` () =
+  let result =
+    {
+      Columns = [ { Name = "name"; Type = StringType } ]
+      Rows = [ [| String "\"John Quoted\"" |] ]
+    }
+
+  let suppressedAnonCount = Null
+  let CSVResult = (CSVFormatter.format suppressedAnonCount result).Split('\n')
+
+  CSVResult.[1] |> should equal "\"\"\"John Quoted\"\"\""
+  CSVResult.Length |> should equal 2
+
+[<Fact>]
+let ``Formats nulls`` () =
+  let result =
+    {
+      Columns =
+        [
+          { Name = "age"; Type = IntegerType }
+          { Name = "city"; Type = StringType }
+          { Name = "discount"; Type = RealType }
+          { Name = "active"; Type = BooleanType }
+          { Name = "count"; Type = IntegerType }
+        ]
+      Rows = [ [| Null; Null; Null; Null; Null |] ]
+    }
+
+  let suppressedAnonCount = Null
+  let CSVResult = (CSVFormatter.format suppressedAnonCount result).Split('\n')
+
+  CSVResult.[1] |> should equal "NULL,NULL,NULL,NULL,NULL"
```

Comment:
Aren't we supposed to treat `NULL` as empty string?

Reply:
good question, but it seems it's been `NULL` for a long time, c.f. 
https://github.com/diffix/desktop/blob/master/anonymizer/OpenDiffix.Service/Program.fs#L36
and https://github.com/diffix/reference/blob/master/src/OpenDiffix.Core/Value.fs#L14

Reply:
That looks like a pseudo-csv that we have been emitting. We may need to represent `NULL` as `""`. @cristianberneanu?

Reply:
> when I plant some NULLs (,,) into the input CSV

CSV format doesn't support NULL fields. The parser probably converts those to "" automatically.
The Desktop export query should never generate NULL values in the output, so we shouldn't worry about it.

Reply:
Sorry about raising false alert on this, and thank you for raising the issue. I'm scratching my head here - if I inject a `printfn` into `CSVFormatter` result and run it, I'm getting different results in the `printfn`, depending on how I run it:
1/ `NULL` if I run the `CSVFormatterTests`
2/ `""` if I run `OpenDiffix.Service` `Export` request manually

I'm going to investigate why this happens, I'm sure the problem lies with the tests, but in some obscure manner.

Reply:
Like Cristian said. The output of the csv reader has no nulls, whereas your tests may have them specifically set.

Reply:
I am able to export NULLs from `desktop v1.0.1` if I use generalization on the integer column. Also, even without the nulls in the input data, we provide inconsistent output data in generalized / non-generalized request:

```
"passenger_count","count"
2,67315
0,324384
4,33573
6,15004
```

```
"passenger_count","count"
"2",52441
"1",324332
"4",6345
"3",14917
"5",27193
"6",15007
```

Generalization causes the query to `cast(passenger_count as real)` which removes the quote and enables the export to produce `NULL`, if any are present in the input